cmd/consul-dataplane/config.go

```diff
@@ -0,0 +1,315 @@
+}
+
+type PrometheusTelemetryFlags struct {
+	RetentionTime     *Duration `json:"retentionTime,omitempty"`
+	CACertsPath       *string   `json:"caCertsPath,omitempty"`
+	KeyFile           *string   `json:"keyFile,omitempty"`
+	CertFile          *string   `json:"certFile,omitempty"`
+	ServiceMetricsURL *string   `json:"serviceMetricsURL,omitempty"`
+	ScrapePath        *string   `json:"scrapePath,omitempty"`
+	MergePort         *int      `json:"mergePort,omitempty"`
+}
+
+type EnvoyFlags struct {
+	AdminBindAddr    *string `json:"adminBindAddress,omitempty"`
+	AdminBindPort    *int    `json:"adminBindPort,omitempty"`
+	ReadyBindAddr    *string `json:"readyBindAddress,omitempty"`
+	ReadyBindPort    *int    `json:"readyBindPort,omitempty"`
+	Concurrency      *int    `json:"concurrency,omitempty"`
+	DrainTimeSeconds *int    `json:"drainTimeSeconds,omitempty"`
+	DrainStrategy    *string `json:"drainStrategy,omitempty"`
+
+	ShutdownDrainListenersEnabled *bool   `json:"shutdownDrainListenersEnabled,omitempty"`
+	ShutdownGracePeriodSeconds    *int    `json:"shutdownGracePeriodSeconds,omitempty"`
+	GracefulShutdownPath          *string `json:"gracefulShutdownPath,omitempty"`
+	GracefulPort                  *int    `json:"gracefulPort,omitempty"`
+	DumpEnvoyConfigOnExitEnabled  *bool   `json:"dumpEnvoyConfigOnExitEnabled,omitempty"`
+}
+
+const (
+	DefaultLogName = "consul-dataplane"
+)
+
+// buildDataplaneConfig builds the necessary config needed for the
+// dataplane to start. We begin with the default version of the dataplane
+// config(with the default values) followed by merging the file based
+// config generated from the `-config-file` input into it.
+// Since values given via CLI flags take the most precedence, we finally
+// merge the config generated from the flags into the previously
+// generated/merged config
+func (f *FlagOpts) buildDataplaneConfig(extraArgs []string) (*consuldp.Config, error) {
+	consulDPDefaultFlags, err := buildDefaultConsulDPFlags()
+	if err != nil {
+		return nil, err
+	}
+
+	if f.configFile != "" {
+		consulDPFileBasedFlags, err := f.buildConfigFromFile()
+		if err != nil {
+			return nil, err
+		}
+
+		mergeConfigs(consulDPDefaultFlags, consulDPFileBasedFlags)
+	}
+
+	consulDPCLIFlags := &f.dataplaneConfig
+	mergeConfigs(consulDPDefaultFlags, consulDPCLIFlags)
+
+	consuldpRuntimeConfig, err := constructRuntimeConfig(consulDPDefaultFlags, extraArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	return consuldpRuntimeConfig, nil
+}
+
+// Constructs a config based on the values present in the config json file
+func (f *FlagOpts) buildConfigFromFile() (*DataplaneConfigFlags, error) {
+	var cfg *DataplaneConfigFlags
+	data, err := os.ReadFile(f.configFile)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(data, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// Constructs a config with the default values
+func buildDefaultConsulDPFlags() (*DataplaneConfigFlags, error) {
+	data := `
```

Comment:
Since the fields of `DataplaneConfigFlags` are mostly pointers, it kind of looks ugly referencing the const variables inside the struct. We can't define `const` variables for these default flags and assign them as `&DefineConsulFlag` to those fields. We would have to fallback to primitive types for these variables which doesn't look that good as well. So I went ahead with a hardcoded json that is easier for people to make changes. I can reiterate on this if people see further concerns.